File: 백준/Silver/1388. 바닥 장식/바닥 장식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int count = 0;

        String[][] arr = new String[N][M];
        for (int i = 0; i < N; i++) {
            arr[i] = br.readLine().split("");
        }

        for (int r = 0; r < N; r++) {
            for (int c = 0; c < M; c++) {
                if (arr[r][c].equals("-")) {
                    dfs_row(r, c, M, arr);
                    count++;
                } else if (arr[r][c].equals("|")) {
                    dfs_col(r, c, N, arr);
                    count++;
                }
            }
        }
        System.out.println(count);

    }

    public static void dfs_row(int r, int  c, int M, String[][]arr) {
        arr[r][c] = ".";
        List<int[]> S = new ArrayList<>();
        S.add(new int[]{r, c});

        while (!S.isEmpty()) {
            int[] s = S.remove(S.size() - 1);
            r = s[0];
            c = s[1];

            for (int dc : new int[]{-1, 1}) {
                int nc = c + dc;
                if ((nc < 0) || (nc >= M)) continue;
                if (arr[r][nc].equals("-")) {
                    arr[r][nc] = ".";
                    S.add(new int[]{r, nc});
                }
            }
        }
    }

    public static void dfs_col(int r, int  c, int N, String[][]arr) {
        arr[r][c] = ".";
        List<int[]> S = new ArrayList<>();
        S.add(new int[]{r, c});

        while (!S.isEmpty()) {
            int[] s = S.remove(S.size() - 1);
            r = s[0];
            c = s[1];

            for (int dr : new int[]{-1, 1}) {
                int nr = r + dr;
                if ((nr < 0) || (nr >= N)) continue;
                if (arr[nr][c].equals("|")) {
                    arr[nr][c] = ".";
                    S.add(new int[]{nr, c});
                }
            }
        }
    }
}

File: 백준/Silver/1388. 바닥 장식/바닥 장식.py
def dfs_row(r, c):
    global N, M
    arr[r][c] = '.'
    S = [(r, c)]

    while S:
        r, c = S.pop()

        for dc in (-1, 1):
            nc = c + dc
            if nc < 0 or nc >= M: continue
            if arr[r][nc] == '-':
                arr[r][nc] = '.'
                S.append((r, nc))

def dfs_col(r, c):
    global N, M
    arr[r][c] = '.'
    S = [(r, c)]

    while S:
        r, c = S.pop()

        for dr in (-1, 1):
            nr = r + dr
            if nr < 0 or nr >= N: continue
            if arr[nr][c] == '|':
                arr[nr][c] = '.'
                S.append((nr, c))


N, M = map(int, input().strip().split())    # 행, 열
arr = [list(input()) for _ in range(N)]

cnt = 0
for r in range(N):
    for c in range(M):
        if arr[r][c] == '-':
            dfs_row(r, c)
            cnt += 1
        elif arr[r][c]  == '|':
            dfs_col(r, c)
            cnt += 1
print(cnt)
